pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+		return false, fmt.Errorf("abandoning command, reached timeout, %w", cmd.LastError)
+	}
+	// If the time hasn't expired, either wait, or terminate.
+	requeue, err := q.WaitOrTerminate(ctx, cmd)
+	if err != nil {
+		// If there was an error, set this as the command's last error so that we can propagate it.
+		cmd.LastError = err
+		if !retry.IsRecoverable(err) {
+			return false, nil
+		}
+		return requeue, nil
+	}
+	return requeue, nil
+}
+
+// WaitOrTerminate will wait until launched machines are ready.
+// Once the replacements are ready, it will terminate the candidates.
+// Will return true if the item in the queue should be re-queued. If a command has
+// timed out, this will return false.
+// nolint:gocyclo
+func (q *Queue) WaitOrTerminate(ctx context.Context, cmd *Command) (bool, error) {
+	waitErrs := make([]error, len(cmd.ReplacementKeys))
+	for i := range cmd.ReplacementKeys {
+		key := cmd.ReplacementKeys[i]
+		// If we know the node claim is initialized, no need to check again.
+		if key.Initialized {
+			continue
+		}
+		// Get the nodeclaim
+		nodeClaim, err := nodeclaim.Get(ctx, q.kubeClient, key.Key)
+		if err != nil {
+			// The NodeClaim got deleted after an initial eventual consistency delay
+			// This means that there was an ICE error or the Node initializationTTL expired
+			// In this case, the error is unrecoverable, so don't requeue.
+			if errors.IsNotFound(err) && q.clock.Since(cmd.TimeAdded) > time.Second*5 {
+				return false, retry.Unrecoverable(fmt.Errorf("getting machine, %w", err))
+			}
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		// We emitted this event when Deprovisioning was blocked on launching/termination.
+		// This does not block other forms of deprovisioning, but we should still emit this.
+		q.recorder.Publish(deprovisioningevents.Launching(nodeClaim, cmd.Reason))
+		if !nodeClaim.StatusConditions().GetCondition(v1beta1.Initialized).IsTrue() {
+			q.recorder.Publish(deprovisioningevents.WaitingOnReadiness(nodeClaim))
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		key.Initialized = true
+		// This should only be reached once since initialization is checked at the beginning.
+		deprovisioningReplacementNodeInitializedHistogram.Observe(q.clock.Since(cmd.TimeAdded).Seconds())
+	}
+	// If we have any errors, don't continue
+	if err := multierr.Combine(waitErrs...); err != nil {
+		return true, err
+	}
+
+	// Reaching here means we know that all replacements have been provisioned.
+	// All we need to do now is get a successful delete call for each node claim,
+	// then the termination controller will handle the eventual deletion of the nodes.
+	errs := make([]error, len(cmd.Candidates))
+	workqueue.ParallelizeUntil(ctx, len(cmd.Candidates), len(cmd.Candidates), func(i int) {
+		q.recorder.Publish(deprovisioningevents.Terminating(cmd.Candidates[i].Node, cmd.Candidates[i].NodeClaim, cmd.Reason)...)
+		errs[i] = retry.Do(func() error {
```

Comment:
It feels wrong to me to retry in this requeue-able action here. What if we just throw an error and let our retry logic handle this?

Reply:
Was going back and forth on this one. We can either do the requeue-able parallelizeUntil here (where we know that all replacements have initialized, so we just need the successful delete calls), or we can change the `timeAdded` field and just re-enqueue and have a 10 minute timeout on a successful termination call. I could go either way. Seems like you have a preference to re-enqueue and wait for the item to be popped off again?